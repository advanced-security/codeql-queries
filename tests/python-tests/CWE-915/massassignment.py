from flask import Flask, request, abort

app = Flask(__name__)


class Users(object):
    uid: int
    username: str
    password: str


@app.route("/test1")
def test1():
    # Test 1: Simple test case
    i = request.args.get("i")
    user = Users()
    
    # codeql: py/mass-assignment
    setattr(user, request.args.get("k"), request.args.get("v"))
    
    # codeql: py/mass-assignment
    user.__setattr__(request.args.get("k"), request.args.get("v"))

    # false-positive: variable isn't user controlled
    setattr(user, "uid", request.args.get("uid"))
    # same as: user.uid = i
    
    # false-positive: variable isn't user controlled
    user.__setattr__("uid", request.args.get("uid"))
    # same as: user.uid = i
    
    return f"<h1>Test 3</h1>"


@app.route("/test3")
def test3():
    user = Users()
    # Test 2:
    # remote-flow: requrst.args
    for param_name, param_vaule in request.args.items():
        # codeql: py/mass-assignment
        setattr(user, param_name, param_vaule)
    
    return f"<h1>Test 3</h1>"


@app.route("/test4")
def test4():
    # Test 3:
    if request.method == "PUT":
        if not request.json:
                abort(400)

        user_data = request.json
        user = Users()

        # remote-flow: requrst.args
        for item in user_data:
            # codeql: py/mass-assignment
            setattr(user, item, request.json[item])

    return f"<h1>Test 4</h1>"
