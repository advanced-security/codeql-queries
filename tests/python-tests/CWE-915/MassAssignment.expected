edges
| massassignment.py:15:9:15:15 | ControlFlowNode for request | massassignment.py:19:19:19:30 | ControlFlowNode for Attribute |
| massassignment.py:19:19:19:25 | ControlFlowNode for request | massassignment.py:19:19:19:30 | ControlFlowNode for Attribute |
| massassignment.py:19:19:19:30 | ControlFlowNode for Attribute | massassignment.py:19:19:19:39 | ControlFlowNode for Attribute() |
| massassignment.py:40:9:40:18 | SSA variable param_name | massassignment.py:42:23:42:32 | ControlFlowNode for param_name |
| massassignment.py:40:9:40:31 | ControlFlowNode for Tuple | massassignment.py:40:9:40:18 | SSA variable param_name |
| massassignment.py:40:9:40:31 | ControlFlowNode for Tuple [Tuple element at index 0] | massassignment.py:40:9:40:18 | SSA variable param_name |
| massassignment.py:40:9:40:31 | IterableElement | massassignment.py:40:9:40:31 | ControlFlowNode for Tuple |
| massassignment.py:40:9:40:31 | IterableElement | massassignment.py:40:9:40:31 | ControlFlowNode for Tuple [Tuple element at index 0] |
| massassignment.py:40:9:40:31 | IterableSequence | massassignment.py:40:9:40:31 | ControlFlowNode for Tuple |
| massassignment.py:40:9:40:31 | IterableSequence | massassignment.py:40:9:40:31 | IterableElement |
| massassignment.py:40:36:40:42 | ControlFlowNode for request | massassignment.py:40:36:40:47 | ControlFlowNode for Attribute |
| massassignment.py:40:36:40:47 | ControlFlowNode for Attribute | massassignment.py:40:9:40:31 | IterableSequence |
| massassignment.py:50:8:50:14 | ControlFlowNode for request | massassignment.py:54:21:54:32 | ControlFlowNode for Attribute |
| massassignment.py:51:16:51:22 | ControlFlowNode for request | massassignment.py:54:21:54:32 | ControlFlowNode for Attribute |
| massassignment.py:54:21:54:27 | ControlFlowNode for request | massassignment.py:54:21:54:32 | ControlFlowNode for Attribute |
| massassignment.py:54:21:54:32 | ControlFlowNode for Attribute | massassignment.py:58:13:58:16 | SSA variable item |
| massassignment.py:58:13:58:16 | SSA variable item | massassignment.py:60:27:60:30 | ControlFlowNode for item |
nodes
| massassignment.py:15:9:15:15 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:19:19:19:25 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:19:19:19:30 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| massassignment.py:19:19:19:39 | ControlFlowNode for Attribute() | semmle.label | ControlFlowNode for Attribute() |
| massassignment.py:40:9:40:18 | SSA variable param_name | semmle.label | SSA variable param_name |
| massassignment.py:40:9:40:31 | ControlFlowNode for Tuple | semmle.label | ControlFlowNode for Tuple |
| massassignment.py:40:9:40:31 | ControlFlowNode for Tuple [Tuple element at index 0] | semmle.label | ControlFlowNode for Tuple [Tuple element at index 0] |
| massassignment.py:40:9:40:31 | IterableElement | semmle.label | IterableElement |
| massassignment.py:40:9:40:31 | IterableSequence | semmle.label | IterableSequence |
| massassignment.py:40:36:40:42 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:40:36:40:47 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| massassignment.py:42:23:42:32 | ControlFlowNode for param_name | semmle.label | ControlFlowNode for param_name |
| massassignment.py:50:8:50:14 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:51:16:51:22 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:54:21:54:27 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:54:21:54:32 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| massassignment.py:58:13:58:16 | SSA variable item | semmle.label | SSA variable item |
| massassignment.py:60:27:60:30 | ControlFlowNode for item | semmle.label | ControlFlowNode for item |
subpaths
#select
| massassignment.py:19:19:19:39 | ControlFlowNode for Attribute() | massassignment.py:15:9:15:15 | ControlFlowNode for request | massassignment.py:19:19:19:39 | ControlFlowNode for Attribute() | Use of $@. | massassignment.py:15:9:15:15 | ControlFlowNode for request | mass assignment |
| massassignment.py:19:19:19:39 | ControlFlowNode for Attribute() | massassignment.py:19:19:19:25 | ControlFlowNode for request | massassignment.py:19:19:19:39 | ControlFlowNode for Attribute() | Use of $@. | massassignment.py:19:19:19:25 | ControlFlowNode for request | mass assignment |
| massassignment.py:42:23:42:32 | ControlFlowNode for param_name | massassignment.py:40:36:40:42 | ControlFlowNode for request | massassignment.py:42:23:42:32 | ControlFlowNode for param_name | Use of $@. | massassignment.py:40:36:40:42 | ControlFlowNode for request | mass assignment |
| massassignment.py:60:27:60:30 | ControlFlowNode for item | massassignment.py:50:8:50:14 | ControlFlowNode for request | massassignment.py:60:27:60:30 | ControlFlowNode for item | Use of $@. | massassignment.py:50:8:50:14 | ControlFlowNode for request | mass assignment |
| massassignment.py:60:27:60:30 | ControlFlowNode for item | massassignment.py:51:16:51:22 | ControlFlowNode for request | massassignment.py:60:27:60:30 | ControlFlowNode for item | Use of $@. | massassignment.py:51:16:51:22 | ControlFlowNode for request | mass assignment |
| massassignment.py:60:27:60:30 | ControlFlowNode for item | massassignment.py:54:21:54:27 | ControlFlowNode for request | massassignment.py:60:27:60:30 | ControlFlowNode for item | Use of $@. | massassignment.py:54:21:54:27 | ControlFlowNode for request | mass assignment |
